okio/src/commonMain/kotlin/okio/internal/Buffer.kt

```diff
@@ -565,6 +608,124 @@ internal inline fun Buffer.commonRead(sink: ByteArray, offset: Int, byteCount: I
   return toCopy
 }
 
+internal inline fun Buffer.commonReadDecimalLong(): Long {
+  if (size == 0L) throw EOFException()
+
+  // This value is always built negatively in order to accommodate Long.MIN_VALUE.
+  var value = 0L
+  var seen = 0
+  var negative = false
+  var done = false
+
+  val overflowZone = Long.MIN_VALUE / 10L
+  var overflowDigit = Long.MIN_VALUE % 10L + 1
+
+  do {
+    val segment = head!!
+
+    val data = segment.data
+    var pos = segment.pos
+    val limit = segment.limit
+
+    while (pos < limit) {
+      val b = data[pos]
+      if (b >= '0'.toByte() && b <= '9'.toByte()) {
+        val digit = '0'.toByte() - b
+
+        // Detect when the digit would cause an overflow.
+        if (value < overflowZone || value == overflowZone && digit < overflowDigit) {
+          val buffer = Buffer().writeDecimalLong(value).writeByte(b.toInt())
+          if (!negative) buffer.readByte() // Skip negative sign.
+          throw NumberFormatException("Number too large: ${buffer.readUtf8()}")
+        }
+        value *= 10L
+        value += digit.toLong()
+      } else if (b == '-'.toByte() && seen == 0) {
+        negative = true
+        overflowDigit -= 1
+      } else {
+        if (seen == 0) {
+          throw NumberFormatException(
+            "Expected leading [0-9] or '-' character but was ${b.toHexString()}")
+        }
+        // Set a flag to stop iteration. We still need to run through segment updating below.
+        done = true
+        break
+      }
+      pos++
+      seen++
+    }
+
+    if (pos == limit) {
+      head = segment.pop()
+      SegmentPool.recycle(segment)
+    } else {
+      segment.pos = pos
+    }
+  } while (!done && head != null)
+
+  size -= seen.toLong()
+  return if (negative) value else -value
+}
+
+internal inline fun Buffer.commonReadHexadecimalUnsignedLong(): Long {
+  if (size == 0L) throw EOFException()
+
+  var value = 0L
+  var seen = 0
+  var done = false
+
+  do {
+    val segment = head!!
+
+    val data = segment.data
+    var pos = segment.pos
+    val limit = segment.limit
+
+    while (pos < limit) {
+      val digit: Int
+
+      val b = data[pos]
+      if (b >= '0'.toByte() && b <= '9'.toByte()) {
+        digit = b - '0'.toByte()
+      } else if (b >= 'a'.toByte() && b <= 'f'.toByte()) {
+        digit = b - 'a'.toByte() + 10
+      } else if (b >= 'A'.toByte() && b <= 'F'.toByte()) {
+        digit = b - 'A'.toByte() + 10 // We never write uppercase, but we support reading it.
+      } else {
+        if (seen == 0) {
+          throw NumberFormatException(
+            "Expected leading [0-9a-fA-F] character but was ${b.toHexString()}")
+        }
+        // Set a flag to stop iteration. We still need to run through segment updating below.
+        done = true
+        break
+      }
+
+      // Detect when the shift will overflow.
```

Comment:
interesting, both  this and the decimal case support an arbitrary number of leading `0` bytes

Reply:
I’m sure I already knew that, but now I’m relearning it!